src/bin/ion/commands/structural_recursion.rs

```diff
@@ -0,0 +1,120 @@
+use anyhow::Result;
+use ion_rs::{AnyEncoding, Element, IonType, LazyValue, ValueRef};
+
+/// Trait for operations that transform Ion elements,
+/// used for transformations like timestamp conversions
+pub trait ElementMapper {
+    fn map(&self, element: Element) -> Result<Element>;
+}
+
+/// Trait for operations that analyze Ion values without transformation,
+/// used for analysis like depth calculation that only need to examine values
+pub trait ValueVisitor<T> {
+    fn visit(&mut self, value: ValueRef<AnyEncoding>, depth: usize) -> Result<()>;
+    fn result(self) -> T;
+}
+
+/// Iteratively applies a mapper to all elements in an Ion structure
+/// This function uses an explicit work stack instead of recursion
+/// It processes the structure in post-order (children before parents) to enable reconstruction
```

Comment:
Post-order makes it more difficult to create a mapper which efficiently drops or mutates or members. Would it be better to map each value and then descend into the result?

As an aside, this also makes me wonder [ControlFlow](https://doc.rust-lang.org/beta/std/ops/enum.ControlFlow.html) could be useful for signaling when the mapper knows that further recursion is not needed. I'm not suggesting that you use ControlFlow now- we have only two extant use cases and both depend on total exploration of the value.